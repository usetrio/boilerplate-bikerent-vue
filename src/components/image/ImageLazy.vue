<script lang="ts">
/* global Maybe */
import { defineComponent, nextTick } from 'vue'
import { debounce } from 'lodash-es'

interface ImageLazyData {
  observer: Maybe<IntersectionObserver>
}

export default defineComponent({
  name: 'ImageLazy',
  props: {
    src: {
      type: String,
      required: true
    }
  },
  data: (): ImageLazyData => ({
    observer: null
  }),
  computed: {
    image(): Maybe<HTMLImageElement> {
      return this.$refs.image as HTMLImageElement
    }
  },
  watch: {
    async src() {
      this.clearImage()
      if (!this.src) {
        return
      }

      await nextTick()

      this.initObserver()
    }
  },
  beforeMount() {
    this.observer = new IntersectionObserver(
      debounce(async ([entry]) => {
        const { isIntersecting } = entry

        if (isIntersecting && !this.image?.src) {
          this.setImage()

          await nextTick()
          this.observer!.disconnect()
        }
      }, 100),
      {
        threshold: 0
      }
    )
  },
  async mounted() {
    this.clearImage()
    await nextTick()

    this.initObserver()
  },
  beforeUnmount() {
    this.observer?.disconnect()
  },
  methods: {
    initObserver() {
      this.observer?.disconnect()
      this.image && this.observer?.observe(this.image)
    },
    setImage() {
      if (!this.image) {
        return
      }

      this.image.src = this.src
    },
    clearImage() {
      if (!this.image) {
        return
      }

      this.image.removeAttribute('src')
    }
  }
})
</script>

<template>
  <img ref="image" class="image image--lazy" />
</template>

<style lang="scss" scoped>
.image-lazy {
  position: relative;
}
</style>
