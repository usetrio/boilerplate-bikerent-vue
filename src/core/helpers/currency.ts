/**
 * Formats value to currency notation
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat
 *
 * @param {number} value
 * @param {string} currency Currency Acronym (EUR, USD, BRL)
 * @returns {string}
 */
export function formatCurrency(value: number, currency: string): string {
  const formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
    currencyDisplay: 'symbol',
    maximumFractionDigits: value % 1 > 0 ? 2 : 0
  })

  const parts = formatter.formatToParts(value)

  const currencyLabel = parts.find(({ type }) => type === 'currency')!.value
  const formattedValue = parts
    .filter(({ type }) => type !== 'currency')
    .map(({ value }) => value)
    .reduce((string, part) => string + part)

  return `${formattedValue} ${currencyLabel}`
}
