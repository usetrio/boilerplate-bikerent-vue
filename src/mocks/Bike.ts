import type { BikeRentDetails, BikeReturnDetails } from '@/core/api/modules/bike'

const [mockDate] = new Date().toISOString().split('T')

const mockedImageUrls = [
  'https://cremecycles.com/images/glowne/15.jpg',
  'https://cremecycles.com/images/glowne/13.jpg',
  'https://cremecycles.com/images/glowne/14.jpg'
]

const mockedBike: Bike = {
  id: 1,
  candidateId: 5,
  name: 'Murazik, Thiel and Robel',
  type: 'Cyclocross Bicycle',
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
  bodySize: 24,
  city: 'Manhattan',
  cardImage: 'https://cremecycles.com/images/glowne/15.jpg',
  imageUrls: mockedImageUrls,
  maxLoad: 110,
  rate: 135,
  ratings: 4.8
}

const mockBikeRentDetails: BikeRentDetails = {
  userId: 1,
  bikeId: mockedBike.id,
  dateFrom: mockDate,
  dateTo: mockDate
}

const mockedBikeReturnDetails: BikeReturnDetails = {
  userId: 1,
  bikeId: mockedBike.id
}

const mockedBikeList = [mockedBike, mockedBike, mockedBike, mockedBike]

export { mockedBike, mockedBikeList, mockedImageUrls, mockBikeRentDetails, mockedBikeReturnDetails }
