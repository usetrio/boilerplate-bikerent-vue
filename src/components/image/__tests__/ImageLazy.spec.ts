import { describe, expect, it, vi } from 'vitest'
import { flushPromises, mount } from '@vue/test-utils'

import { ImageLazy } from '@/components/image'
import { mockIntersectionObserver } from '@/core/helpers/tests'
import { mockedImageUrls } from '@/mocks'

describe('ImageLazy', () => {
  const [image1, image2, image3] = mockedImageUrls

  vi.stubGlobal('IntersectionObserver', mockIntersectionObserver)
  vi.mock('lodash-es', () => ({
    debounce: vi.fn((fn) => fn)
  }))

  it('renders correctly', () => {
    const wrapper = mount(ImageLazy, { props: { src: image1 } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('render image element', () => {
    const wrapper = mount(ImageLazy, { props: { src: image2 } })

    expect(wrapper.find('img').exists()).toBe(true)
  })

  it('renders image src', async () => {
    const wrapper = mount(ImageLazy, { props: { src: image3 } })

    await flushPromises()

    expect(wrapper.find('img').attributes('src')).equals(image3)
  })

  it('renders empty image', async () => {
    const wrapper = mount(ImageLazy, { props: { src: '' } })

    await flushPromises()

    expect(wrapper.find('img').attributes('src')).toBe('')
  })

  it('renders different image', async () => {
    const wrapper = mount(ImageLazy, { props: { src: image3 } })

    await flushPromises()

    expect(wrapper.find('img').attributes('src')).equals(image3)

    wrapper.setProps({ src: image1 })
    await flushPromises()

    expect(image1, wrapper.find('img').attributes('src')).equals(image1)
  })
})
