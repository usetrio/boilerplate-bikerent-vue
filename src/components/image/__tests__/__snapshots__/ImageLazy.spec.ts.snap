// Vitest Snapshot v1

exports[`ImageLazy > renders correctly 1`] = `"<img class=\\"image image--lazy\\" data-v-ff3f7011=\\"\\">"`;
